### PR TITLE
Support dynamic setting of DropBlock.dropout_rate #176

### PR DESCRIPTION
**Pull request by kksinghal**
Three options for users:
1. Pass dropout_rate for constant dropout
2. Pass initial_dropout_rate, end_dropout_rate, total_train_steps. We handle the linear scheduler through function _linear_scheduler implemented within DropBlock2D class
3. Pass a custom function, which takes current_train_step as input, and returns a float dropout_rate

### CODE REVIEW

**Comment by LukeWood** [comment]
Lets do something like https://github.com/keras-team/keras-cv/pull/238 but for schedules

**Comment by kksinghal** [comment]
@LukeWood Please have a look

**Comment by LukeWood** [comment]
@kksinghal Thanks for the API.  (Sorry this got lost in my inbox).  

Due to the "core" nature of this API I am still figuring out some details as to how this will fit into the entire ecosystem.  I think your design is good, I just want to be sure it is "future proof" before I accept and merge.  But overall this PR looks pretty good.  This is something I will be looking at soon now that v1 is out.

**Comment by lucasdavid** [comment]
Just a few thoughts on this, if it's of any help:

1. There's a leak of responsibility, where the Dropblock is now also expected to keep track of training status.
2. The paper clearly states that all Dropblock layers receive the same alpha, so having different schedulers and `training_step` for each layer is outside the scope of this technique and hasn't been validated by any scientific work, as far as I know.
1. The method `DropBlock2D#call` will not be invoked each feed-forward if its traced. Instead, this function is converted into a TF graph, which will bake the number `1` as a constant and always return the exact same value. You need to use a tf.Variable in here and update it each step/epoch. There's already a variable available for this (`model.optimizer.iterations <tf.Tensor[type: int, shape: ()]>`). Maybe there's a way to reuse it in here?

**Comment by sebastian-sz** [comment]
I'm wondering if wouldn't it be better to write a separate, lightweight wrapper layer `ScheduledDropBlock2D`, (could even be in the same file) which accepts scheduler + all those other params and calls regular `DropBlock2D` internally.

This would simplify the constructor and maintain the separation of concerns.

**Comment by LukeWood** [comment]
> I'm wondering if wouldn't it be better to write a separate, lightweight wrapper layer `ScheduledDropBlock2D`, (could even be in the same file) which accepts scheduler + all those other params and calls regular `DropBlock2D` internally.
> 
> This would simplify the constructor and maintain the separation of concerns.

We're also still figuring out what the best way to design this API is.  There are valid use cases in Keras core as well.
